os/pm/pm_ioctl.c

```diff
@@ -0,0 +1,143 @@
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: pm_ioctl
+ *
+ * Description:
+ *   This function can be called from applications running on TizenRT to
+ *   request/enable/disable power state transitions. This function conveys
+ *   the request from the application to the PM module, which thereafter
+ *   checks the feasibility of the requests. There is no returned value.
+ *
+ *   The power management state is not automatically changed, however.
+ *   up_idlepm() must be called in order to make the state change.
+ *
+ *   These two steps are separated because the application is not supposed
+ *   to invoke power state changes directly, it can only request/suggest.
+ *
+ * Input Parameters:
+ *   domain - the PM domain to check
+ *
+ * Returned Value:
+ *   None
+ *
+ ****************************************************************************/
+
+void pm_ioctl(enum pm_ioctl_cmds cmd)
+{
+	switch(cmd) {
+		case PM_IOC_NOCMD:
+			dbg("Nothing to do\n");
+			break;
+		case PM_IOC_STAY:
+			//Get the current domain
+			//perform pm_stay to block the domain in its current PM state
+			//pm_stay(<DOMAIN>, pm_querystate(<DOMAIN>));
+			break;
+		case PM_IOC_RELAX:
+			//Get the current domain
+			//perform pm_relax to unblock the domain's state transitions
+			//pm_relax(<DOMAIN>, pm_querystate(<DOMAIN>));
+			break;
+		/* Note that in the above 2 IOCTL calls, the application does not provide the state to stay/relax.
+		 * The application cannot anyways know the pm_domain of the current thread.
+		 * The thread will be extracted during the processing of the IOCTL call itself.
+		 * The state to stay in is the current state itself, and the state to be relaxed from will also
+		 * be the current state itself. This is based on the assumption that if an application is requesting
+		 * a stay, then the current state is capable enough for the application to do its work, hence we will
+		 * stay in the current request only. If a stay was invoked in the current PM state, then it should be
+		 * relaxed from the same state too.
+		 * Open to discussion and consideration
+		 */
+		case PM_IOC_SLEEP:
+			//Get the current domain
+			/* Perform any checks neccessary for doing the next step. As per current code in idle.c, some
+			 * preliminary steps are performed in STANDBY state, and then state changes to SLEEP.
+			 */
+			/* Change the recommended state for the current domain to PM_SLEEP
+			 * pdom->recommended = PM_SLEEP
+			 */
+			/* up_idlepm() -> The code in this function may be needed to be modified a bit so that the opening
+			 * condition which invokes a state change is met (oldstate!=newstate)
+			 */
+			break;
+
+			/* If another application running parallely on another core also requests sleep, then the IOCTL
```

Comment:
I think for the current stage, it will be better not to add in secondary core consideration first, as it might be much more complicated then single core scenario I guess?

Reply:
Yes, we can go step by step

Reply:
Just a friendly reminder, if you face any difficulties of reworking the board, please don't hesitate to let us know.
Looking forward to hear your side complete the setup~
Anyway, please share with us once you have some thoughts on the power consumption for current PM structure.